<a href="https://colab.research.google.com/github/MariiaSam/data_science/blob/main/%D0%A2%D0%B5%D0%BC%D0%B0_7_%D0%9A%D0%BB%D0%B0%D1%81%D0%B8%D1%84%D1%96%D0%BA%D0%B0%D1%86%D1%96%D1%8F_%D1%96_%D0%BE%D1%86%D1%96%D0%BD%D0%BA%D0%B0_%D0%BC%D0%BE%D0%B4%D0%B5%D0%BB%D1%96_%D0%A3%D1%80%D0%BE%D0%BA_3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Завдання. Завантажити датасет Student_data.csv . Провести аналіз даних. https://www.kaggle.com/datasets/rabieelkharoua/students-performance-dataset

In [2]:
from google.colab import files

# Завантаження файлу
uploaded = files.upload()

Saving Student_data.csv to Student_data.csv


In [3]:
import pandas as pd

df = pd.read_csv('Student_data.csv')
df

StudentID  Age  Gender  Ethnicity  ParentalEducation  StudyTimeWeekly  \
0          1001   17       1          0                  2        19.833723   
1          1002   18       0          0                  1        15.408756   
2          1003   15       0          2                  3         4.210570   
3          1004   17       1          0                  3        10.028829   
4          1005   17       1          0                  2         4.672495   
...         ...  ...     ...        ...                ...              ...   
2387       3388   18       1          0                  3        10.680555   
2388       3389   17       0          0                  1         7.583217   
2389       3390   16       1          0                  2         6.805500   
2390       3391   16       1          1                  0        12.416653   
2391       3392   16       1          0                  2        17.819907   

      Absences  Tutoring  ParentalSupport  Extracurricular  Sports  Music  \
0            7         1                2                0       0      1   
1            0         0                1                0       0      0   
2           26         0                2                0       0      0   
3           14         0                3                1       0      0   
4           17         1                3                0       0      0   
...        ...       ...              ...              ...     ...    ...   
2387         2         0                4                1       0      0   
2388         4         1                4                0       1      0   
2389        20         0                2                0       0      0   
2390        17         0                2                0       1      1   
2391        13         0                2                0       0      0   

      Volunteering       GPA  GradeClass  
0                0  2.929196         2.0  
1                0  3.042915         1.0  
2                0  0.112602         4.0  
3                0  2.054218         3.0  
4                0  1.288061         4.0  
...            ...       ...         ...  
2387             0  3.455509         0.0  
2388             0  3.279150         4.0  
2389             1  1.142333         2.0  
2390             0  1.803297         1.0  
2391             1  2.140014         1.0  

[2392 rows x 15 columns]

StudentID : унікальний ідентифікатор, призначений кожному студенту (від 1001 до 3392). Демографічні відомості Вік: Вік учнів від 15 до 18 років. Стать : стать студентів, де 0 означає чоловічу, а 1 жіночу. Етнічна приналежність : Етнічна приналежність студентів, закодована таким чином: 0: кавказька 1: афроамериканець 2: азіатський 3: Інше ParentalEducation : рівень освіти батьків, закодований таким чином: 0: Жодного 1: Середня школа 2: Якийсь коледж 3: Бакалавр 4: вище Звички до навчання StudyTimeWeekly : щотижневий час навчання в годинах від 0 до 20. Прогули: кількість пропусків протягом навчального року від 0 до 30. Репетиторство : статус репетиторства, де 0 означає Ні, а 1 означає Так. Участь батьків ParentalSupport : рівень батьківської підтримки, закодований таким чином: 0: Жодного 1: Низький 2: Помірний 3: Високий 4: Дуже високий Позакласна діяльність Позакласний : Участь у позашкільних заходах, де 0 означає Ні, а 1 означає Так. Спорт : участь у спортивних змаганнях, де 0 означає ні, а 1 означає так. Музика : участь у музичних заходах, де 0 означає ні, а 1 означає так. Волонтерська діяльність : участь у волонтерській діяльності, де 0 означає ні, а 1 означає так. Академічна успішність GPA : середній бал за шкалою від 2,0 до 4,0, на який впливають звички до навчання, участь батьків і позакласна діяльність. Цільова змінна: оцінка класу GradeClass : Класифікація оцінок студентів на основі GPA: 0: "A" (середній бал >= 3,5) 1: «B» (3,0 <= GPA < 3,5) 2: «C» (2,5 <= GPA < 3,0) 3: «D» (2,0 <= GPA < 2,5) 4: «F» (середній бал < 2,0)

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2392 entries, 0 to 2391
Data columns (total 15 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   StudentID          2392 non-null   int64  
 1   Age                2392 non-null   int64  
 2   Gender             2392 non-null   int64  
 3   Ethnicity          2392 non-null   int64  
 4   ParentalEducation  2392 non-null   int64  
 5   StudyTimeWeekly    2392 non-null   float64
 6   Absences           2392 non-null   int64  
 7   Tutoring           2392 non-null   int64  
 8   ParentalSupport    2392 non-null   int64  
 9   Extracurricular    2392 non-null   int64  
 10  Sports             2392 non-null   int64  
 11  Music              2392 non-null   int64  
 12  Volunteering       2392 non-null   int64  
 13  GPA                2392 non-null   float64
 14  GradeClass         2392 non-null   float64
dtypes: float64(3), int64(12)
memory usage: 280.4 KB


In [ ]:
df.describe()

StudentID          Age       Gender    Ethnicity  ParentalEducation  \
count  2392.000000  2392.000000  2392.000000  2392.000000        2392.000000   
mean   2196.500000    16.468645     0.510870     0.877508           1.746237   
std     690.655244     1.123798     0.499986     1.028476           1.000411   
min    1001.000000    15.000000     0.000000     0.000000           0.000000   
25%    1598.750000    15.000000     0.000000     0.000000           1.000000   
50%    2196.500000    16.000000     1.000000     0.000000           2.000000   
75%    2794.250000    17.000000     1.000000     2.000000           2.000000   
max    3392.000000    18.000000     1.000000     3.000000           4.000000   

       StudyTimeWeekly     Absences     Tutoring  ParentalSupport  \
count      2392.000000  2392.000000  2392.000000      2392.000000   
mean          9.771992    14.541388     0.301421         2.122074   
std           5.652774     8.467417     0.458971         1.122813   
min           0.001057     0.000000     0.000000         0.000000   
25%           5.043079     7.000000     0.000000         1.000000   
50%           9.705363    15.000000     0.000000         2.000000   
75%          14.408410    22.000000     1.000000         3.000000   
max          19.978094    29.000000     1.000000         4.000000   

       Extracurricular       Sports        Music  Volunteering          GPA  \
count      2392.000000  2392.000000  2392.000000   2392.000000  2392.000000   
mean          0.383361     0.303512     0.196906      0.157191     1.906186   
std           0.486307     0.459870     0.397744      0.364057     0.915156   
min           0.000000     0.000000     0.000000      0.000000     0.000000   
25%           0.000000     0.000000     0.000000      0.000000     1.174803   
50%           0.000000     0.000000     0.000000      0.000000     1.893393   
75%           1.000000     1.000000     0.000000      0.000000     2.622216   
max           1.000000     1.000000     1.000000      1.000000     4.000000   

        GradeClass  
count  2392.000000  
mean      2.983696  
std       1.233908  
min       0.000000  
25%       2.000000  
50%       4.000000  
75%       4.000000  
max       4.000000

In [6]:
df.isnull().sum()

StudentID            0
Age                  0
Gender               0
Ethnicity            0
ParentalEducation    0
StudyTimeWeekly      0
Absences             0
Tutoring             0
ParentalSupport      0
Extracurricular      0
Sports               0
Music                0
Volunteering         0
GPA                  0
GradeClass           0
dtype: int64

In [7]:
df['GradeClass'].value_counts()

GradeClass
4.0    1211
3.0     414
2.0     391
1.0     269
0.0     107
Name: count, dtype: int64

In [8]:
correlation_matrix = df.drop(columns='StudentID').corr()

correlation_with_class = correlation_matrix['GradeClass'].sort_values(ascending=False)

correlation_with_class

GradeClass           1.000000
Absences             0.728633
ParentalEducation    0.041031
Gender               0.022998
Volunteering         0.013156
Age                 -0.006250
Ethnicity           -0.023326
Sports              -0.026654
Music               -0.036065
Extracurricular     -0.069733
Tutoring            -0.111695
StudyTimeWeekly     -0.134131
ParentalSupport     -0.136823
GPA                 -0.782835
Name: GradeClass, dtype: float64


Проаналізувати кількість пропусків для кожного класу успішності

In [9]:
# Групування даних за GradeClass та обчислення середньої кількості пропусків
mean_absences = df.groupby('GradeClass')['Absences'].mean().reset_index()

mean_absences

GradeClass   Absences
0         0.0   5.747664
1         1.0   5.312268
2         2.0   7.250639
3         3.0  11.427536
4         4.0  20.786953

Проаналізувати, як вливає підтримка батьків на середній бал студента GPA

In [10]:
mean_gpa_by_support = df.groupby('ParentalSupport')['GPA'].mean().reset_index()
mean_gpa_by_support

ParentalSupport       GPA
0                0  1.540128
1                1  1.755700
2                2  1.884246
3                3  2.042409
4                4  2.191545

In [11]:
df.duplicated().sum()

0

In [13]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder

In [14]:
df1 = df.drop(columns='StudentID')

# Розбиття на ознаки (X) та цільову змінну (y)
X = df1.drop(columns='GradeClass')
y = df1['GradeClass']

# Перетворення категоріальних змінних (якщо вони є) за допомогою OneHotEncoder
categorical_features = ['Gender', 'Ethnicity', 'ParentalEducation', 'Tutoring', 'ParentalSupport', 'Extracurricular', 'Sports', 'Music', 'Volunteering']
numeric_features = ['Age', 'StudyTimeWeekly', 'Absences', 'GPA']

preprocessor = ColumnTransformer(
    transformers=[
        ('num', StandardScaler(), numeric_features),
        ('cat', OneHotEncoder(), categorical_features)
    ])

# Розбиття даних на тренувальні та тестові набори
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)


RandomForestClassifier


масштабування ознак не обовʼязково


In [19]:
rf_model = Pipeline(steps=[
    ('preprocessor', preprocessor),
    ('classifier', RandomForestClassifier(random_state=42))
])

rf_model.fit(X_train, y_train) # начваємо на навчальних даних

y_pred_rf = rf_model.predict(X_test) # прогнозумємо на тестових даних

accuracy_rf = accuracy_score(y_test, y_pred_rf)
print(f"Точність моделі Random Forest: {accuracy_rf:.4f}")

Точність моделі Random Forest: 0.8997


SVM

In [21]:
svm_model = Pipeline(steps=[
    ('preprocessor', preprocessor),
    ('classifier', SVC())
])

svm_model.fit(X_train, y_train)

y_pred_svm = svm_model.predict(X_test)

accuracy_svm = accuracy_score(y_test, y_pred_svm)
print(f"Точність моделі SVM: {accuracy_svm:.4f}")

Точність моделі SVM: 0.8050


Вивести справжні і прогнозовані значення класів

In [17]:
print("Номер студента | Справжнє значення | Прогнозоване (Random Forest) | Прогнозоване (SVM)")

# Отримання перших 10 студентів
for i, index in enumerate(X_test.index[:10]):
    true_value = y_test.iloc[i]
    pred_rf = y_pred_rf[i]
    pred_svm = y_pred_svm[i]

    # Виведення результатів
    print(f"Студент {index} | Справжнє: {true_value} | RF: {pred_rf} | SVM: {pred_svm}")

Номер студента | Справжнє значення | Прогнозоване (Random Forest) | Прогнозоване (SVM)
Студент 1004 | Справжнє: 4.0 | RF: 4.0 | SVM: 4.0
Студент 196 | Справжнє: 1.0 | RF: 1.0 | SVM: 1.0
Студент 2342 | Справжнє: 2.0 | RF: 3.0 | SVM: 4.0
Студент 1708 | Справжнє: 0.0 | RF: 0.0 | SVM: 1.0
Студент 435 | Справжнє: 4.0 | RF: 4.0 | SVM: 4.0
Студент 321 | Справжнє: 2.0 | RF: 2.0 | SVM: 3.0
Студент 1472 | Справжнє: 3.0 | RF: 3.0 | SVM: 4.0
Студент 1703 | Справжнє: 3.0 | RF: 3.0 | SVM: 3.0
Студент 900 | Справжнє: 3.0 | RF: 3.0 | SVM: 3.0
Студент 1057 | Справжнє: 4.0 | RF: 4.0 | SVM: 4.0
